### PR TITLE
Feature/tree basic

### PR DESCRIPTION
**Pull request by tzahij**
Very raw

---
This change is [Reviewable](https://reviewable.io/reviews/treeverse/lakefs/1013)

### CODE REVIEW

File: forest/tree/tree_reader.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+func (trees *TreesRepoType) NewScanner(treeID gr.TreeID, start gr.Key) (*treeIterator, error) {
+	treeSlice, err := trees.GetTree(treeID)
+	if err != nil {
+		return nil, err
+	}
+	partNum := findPartNumForPath(treeSlice, start)
+	if partNum >= len(treeSlice) {
+		return nil, ErrPathBiggerThanMaxPath
```
Comment (johnnyaug) [inline forest/tree/tree_reader.go:31]: Not sure this is a special case: the next can just return false

File: forest/tree/tree_reader.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+func (t *treeIterator) Close() {
+	if t.currentIter == nil {
+		return
+	}
+	t.currentIter.Close()
+}
+
+func (trees TreesRepoType) GetTree(treeID gr.TreeID) (TreeType, error) {
+	t, exists := trees.TreesMap.Get(string(treeID))
```
Comment (johnnyaug) [inline forest/tree/tree_reader.go:124]: this function should use TierFS in the future, right?

File: forest/tree/tree_reader.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+		t.currentIter.Close()
+		t.closed = true
+		t.err = err
+		return false
+	}
+	return t.currentIter.Next()
+}
+
+func (t *treeIterator) Err() error {
+	if t.currentIter == nil {
```
Comment (johnnyaug) [inline forest/tree/tree_reader.go:103]: I think we're keeping the error on t.err

File: forest/tree/tree_reader.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		currentIter:     partIterator,
+		currentPart:     partNum,
+		treesRepository: trees,
+	}
+	return scanner, nil
+}
+
+func (t *treeIterator) SeekGE(start gr.Key) bool {
+	var err error
+	partNum := findPartNumForPath(t.TreeParts, start)
```
Comment (johnnyaug) [inline forest/tree/tree_reader.go:49]: need to unset t.closed here

File: forest/tree/tree_reader.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	if partNum != t.currentPart {
+		t.currentPart = partNum
```
Comment (johnnyaug) [inline forest/tree/tree_reader.go:51]: If we arrive here after an error, currentIter can be closed

File: forest/tree/tree_writer.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+type TreeWriter struct {
+	activeWriter            sstable.Writer
+	currentPartNumOfEntries int
+	closeAsync              sstable.BatchWriterCloser
+	isSplitKeyFunc          IsSplitKeyFunc
+	splitFactor             uint32
+}
+
+func (tw *TreeWriter) hasOpenWriter() bool {
+	return !(tw.activeWriter == nil)
```
Comment (johnnyaug) [inline forest/tree/tree_writer.go:34]:
```suggestion
	return tw.activeWriter != nil
```

File: forest/tree/tree_writer.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+		return tw.isSplitKeyFunc(key, rowNum)
+	}
+	if tw.splitFactor == 0 {
+		tw.splitFactor = SplitFactor
+	}
+	if uint32(rowNum) >= tw.splitFactor*SplitMaxfactor {
+		return true
+	}
+	fnvHash := fnv.New32a()
+	fnvHash.Write(key)
```
Comment (johnnyaug) [inline forest/tree/tree_writer.go:73]: what is this hash for? is it intended to create parts in different sizes?

File: forest/tree/tree_writer.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+		if err != nil {
+			return err
+		}
+	}
+	return iter.Err()
+}
+
+func (tw *TreeWriter) finalizeTree(reuseParts *TreeType) (gr.TreeID, error) {
+	var newParts TreeType
+	closeResults, err := tw.closeAsync.Wait()
```
Comment (johnnyaug) [inline forest/tree/tree_writer.go:90]: at this point, the activeWriter was already closed?

File: forest/tree/tree_writer.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+		sha.Write([]byte(p.PartName))
+		sha.Write([]byte(p.MaxKey))
+	}
+	hash := sha256.Sum256(nil)
+	treeBytes, err := json.Marshal(tree)
+	if err != nil {
+		return "", err
+	}
+	treeID := gr.TreeID(hex.EncodeToString(hash[:]))
+	err = ioutil.WriteFile("tree_"+string(treeID)+".json", treeBytes, 0666)
```
Comment (itaiad200) [inline forest/tree/tree_writer.go:123]: Pyramid ;) ?

File: forest/tree/tree_writer.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+const (
+	SplitFactor    = 200_000 // avarage number of entries in a part file. used to compute modulo on hash of path
+	SplitMaxfactor = 3       // a part will be closed if number of entries written to it exceed splitFactor * splitMaxFactor
+	SplitMinFactor = 50      // a part will not be closed in number of rows less than splitFactor / SplitMinFactor
+)
+
+type IsSplitKeyFunc func(path gr.Key, rowNum int) bool
+
+type TreeWriter struct {
```
Comment (itaiad200) [inline forest/tree/tree_writer.go:25]: I think the TreeWriter is coupled to `Apply` more than we want it to be.
Some of the inner logic (like flushIterToNewTree) is fine for `Apply`, but might differ for `Merge` for example.

**Review by itaiad200 — CHANGES_REQUESTED** [review]

File: forest/tree/tree_reader.go [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+	//treeID      gr.TreeID
+	TreeParts   TreeType
+	currentIter gr.ValueIterator
+	currentPart int
+	err         error
+	closed      bool
+	trees       *TreesRepoType
+}
+
+func (trees *TreesRepoType) NewScannerFromID(treeID gr.TreeID, start gr.Key) (*treeIterator, error) {
```
Comment (johnnyaug) [inline forest/tree/tree_reader.go:24]: Change "Scanner" to "Iterator"

**Comment by ozkatz** [comment]
Work from this PR was adapted and merged in a different set of PRs. Closing for now.

**Pull request closed without merging**